File: site/Software/POO/EP/ep_cpp/Drone.cpp
#include "Drone.hpp"
#include <iostream>
#include <string>
#include <math.h>

using namespace std;

Drone::Drone(string nome, Bateria* bateria, double posicao){
    this->nome = nome;
    this->bateria = bateria;
    bateria->setUso(true);
    this->posicao = posicao;
    this->altura = 0;
    this->decolado = false;
}

Drone::~Drone(){
    cout << "Drone destruido com sucesso :(" << endl;
}

void Drone::takeoff(int altura){
    decolado = true;
    if(setPosition(posicao, altura)){
        bateria->setCarregavel(false);
        cout << "Takeoff concluido!";
    }
    else
        decolado = false;
}

void Drone::land(){
        altura = 0;
        decolado = false;
        bateria->setCarregavel(true);
        cout << "Drone pousado!" << endl;
}

bool Drone::setPosition(double x, double y){
    if(y <= 0){
        cout << "Drone não pode ir para alturas menores ou iguais a 0" << endl;
        return false;
    }
    int tempo;
    tempo = (sqrt(pow((posicao - x),2) + pow((altura - y),2)))/60;
    if (decolado){
        if(bateria->usar(tempo)){
            posicao = x;
            altura = y;
            cout << "Drone indo para a posicao (" << x << "," << y << ")" << endl;
            cout << "tempo usado " << tempo << endl;
            return true;
        }
        else{
            cout << "Drone nao pode ir ate la" << endl;
            return false;
        }
    }
    else
        cout << "Drone nao esta decolado" << endl;
        return false;
}

double Drone::getPosicao(){
    return posicao;
}

double Drone::getAltura(){
    return altura;
}

int Drone::mapear(){
    cout << "Nao consigo mapear!" << endl;
    return 1;
}

void Drone::status(){
    cout << "Drone: " << nome << endl;
    cout << "Bateria em " << bateria->getCarga() << " mAh e com tempo de voo de " << bateria->calculaTempoDeVoo() << " minutos" << endl;
    cout << "Drone esta na posicao (" << posicao << "," << altura << ")" << endl;
}

File: Software/POO/EP/ep_python/Drone.py
class Drone():
    def __init__(self, nome, bateria, posicao):
        self._nome = nome
        self._bateria = bateria
        self._bateria._uso = True
        self._posicao = posicao
        self._altura = 0.0
        self._decolado = False

    def takeoff(self, altura):
        self._decolado = True
        if self.set_position(self._posicao, altura):
            self._bateria._carregavel = False
            print("Takeoff concluido!\n")
        else:
            self._decolado = False

    def set_position(self, x, y):
        if y <= 0:
            print("\nDrone nao pode ir para menores ou iguais a 0\n")
            return False
        else:
            tempo = (((self._posicao - x)**2 + (self._altura - y)**2)**0.5)/60
            if self._decolado:
                if self._bateria.usar(tempo):
                    self._posicao = x
                    self._altura = y
                    print("\nDrone indo para posicao (" + str(x) + "," + str(y) + ")...")
                    print("Tempo usado: " + str(tempo) +"\n")
                    return True
                return False
            else:
                print("Drone nao esta decolado!")
                return False
            
    def land(self):
        self._altura = 0
        self._decolado = False
        self._bateria._carregavel = True
        print("Drone pousado!\n")

    def mapear(self):
        if self._decolado:
            print("Não consigo mapear!")
        else:
            print("Drone nao esta decolado!")
        return False

    def status(self):
        print("Nome " + self._nome)
        print("Bateria com " + str(self._bateria._carga) + " mAh")
        print("Tempo de Voo restante " + str(self._bateria.calculaTempoDeVoo()))
        print("Drone esta na posicao (" + str(self._posicao) + "," + str(self._altura) + ")")
